sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala

```diff
@@ -36,14 +36,15 @@ import org.apache.spark.util.Utils
 case class CollectLimitExec(limit: Int, child: SparkPlan) extends UnaryExecNode {
   override def output: Seq[Attribute] = child.output
   override def outputPartitioning: Partitioning = SinglePartition
-  override def executeCollect(): Array[InternalRow] = child.executeTake(limit)
-  private val serializer: Serializer = new UnsafeRowSerializer(child.output.size)
+  override def requiredChildDistribution: List[Distribution] = AllTuples :: Nil
+  override def executeCollect(): Array[InternalRow] = {
+    child.collect {
```

Comment:
I think that I understand the intent here: it looks like the idea is to let `EnsureRequirements` handle the shuffling, if necessary, then to strip off the shuffling and walk backwards in the DAG to optimize the `collect()` case. This is pretty clever.

One concern of mine, though, is that there seems to be an implicit assumption on the types of children that this operator can have and I think it would be a good idea to write them down and pattern-match more explicitly. For example, I'm not sure that this strategy is safe in case the child isn't an exchange because then you run the risk of simply dropping any operators that occur post-exchange. I understand that this case can't crop up in the types of plans that we currently generate but it would be good to future-proof by matching explicitly rather than having the less-constrained `collect` behavior here.